### DISCUSSION TITLE
How to set the initial value of the input tag efficiently

### DISCUSSION THREAD

**Discussion by s-ohnishi** [phalcon/cphalcon#15689] [Q&A] [answered]
I want to divide the input fields of the form into meaningful groups
```HTML
<input name = "foo [name]" type = "text">
<input name = "foo [addr]" type = "text">
```
To get this HTML I wrote like this.
```PHP
{{text_field('foo[name]')}}
{{text_field('foo[addr]')}}
```
and
```PHP
$fooValues ​​= $this->request->getPost('foo', array('string','trim'));
```
I think that it is very convenient because it can be obtained as an array and the scope of processing is clarified.

However, to set the value entered using `$this->tag->setDefaults()` in the controller as the initial value
`$this->tag->setDefaults (['foo [name]'=>$fooValues​​['name'], 'foo[addr]'=>$fooValues​​['addr']]);`
must be written.
Is there any other way to rebuild such an array using $fooValues?

I wish I could write `setDefaults(['foo'=>$fooValues])` because I could get it as an array ...

**Comment by Jeckerson** [comment, marked as answer]
Try this one:
```
{{ text_field(['foo[name]', 'value': 'default value']) }}
```

**Reply by s-ohnishi** [reply]
Thank you for your reply.
For example, I want to display "non-fixed values" such as displaying keywords posted on the search form, so I want to set it with the controller.

Is there no choice but to assign the acquired array to the view variable and set it as the default value on the view?

**Reply by quasipickle** [reply]
"foo" doesn't exist as an array anywhere other than when PHP is parsing POST.  Technically speaking, your inputs have the name "foo[name]", with no special attention paid to the square brackets.  Thus, Phalcon doesn't treat it as an array.

You can just run your array through a formatting function to get the correct keys you need:

```php
public function format(string $prefix, array $data): array
{
	$return = [];
	foreach ($data as $key=>$value) {
		$new_key = sprintf("%s[%s]", $prefix, $key);
		$return[$new_key] = $value;
	}
	return $return;
}
```

**Reply by s-ohnishi** [reply]
>"foo" doesn't exist as an array anywhere other than when PHP is parsing POST. 

The data posted by the browser is "foo [name]", but does $_POST['foo'] exist as an array when the PHP script is executed?
Arraying with [ ] as the name attribute value of the input field is a very common method.
And you can also get it as an array with getPost() in Phalcon.

But don't you feel unbalanced if you can't simply set it?
(you can use the array reconstructed using the function you showed us, or you can set it individually on the View)

**Reply by quasipickle** [reply]
`$_POST["foo"]` exists as an array because PHP interprets it as an array.  My point was that technically each field is a separate entity that gets POSTed separately - they don't get POSTed as an array.  Thus, being able to _set_ those fields as a single array is perhaps not matching the spirit of HTTP POSTing.

I agree it might be more convenient in this particular example to be able to set an array.  But you asked how to set the initial value efficiently - one additional function call seems pretty efficient.

**Reply by s-ohnishi** [reply]
I think it's pretty useful to use the function you taught me.
Thank you for that.

I would like to use this to create a wrap function for setDefaults().
I think that it will be supported only for one layer because it becomes complicated and deeper layer is used infrequently.
I asked because it feels like the published samples and manuals don't cover everything.
Guess by reading the arguments of other functions and .zep (actually I can't understand it deeply) I'm proceeding with trials and errors relying on error messages, so I'd be happy if there was disclosure from someone who already had knowledge.

I know that when POSTed data is passed to a web server, it is not an array.
I'm not arguing about that, but under PHP it's an array, and isn't PHP's predecessor irrelevant?
setDefaults() should be able to pass all the post data obtained by getPost () (currently this is a one-dimensional array) as it is.
It is common to get POSTed data as an array already, but I think that it is a very inconvenient specification that it is impossible to pass the obtained data as an array to pass to setDefaults().
